### ISSUE TITLE
Procedure entry point InitOnceBeginInitialize could not be located

### ISSUE THREAD

**Issue by musikBear** [LMMS/lmms#2576]
1.2.0-RC1 ~~1.1.9 RC~~ will not run on xp. ~~That is not an issue, but it need to be in the download info, so peeps with xp does not install in vain.~~
![nokernel32entrypoint](https://cloud.githubusercontent.com/assets/6662066/13109866/581f3b6e-d57c-11e5-855d-c11ae3ef8b7a.jpg)

**Edit, @tresf** quoted exception message to make searchable
> lmms.exe - Entry Point Not Found
> Procedure entry point InitOnceBeginInitialize could not be located in the dynamic link library KERNEL32.dll

A guy with Vista also had problems.
http://lmms.io/forum/viewtopic.php?f=7&t=4822

I will set a xp-exclusive warning on forum. Then at least something has been done.

**Comment by tresf** [comment]
Please don't put up a big flashy warning on the forums.  If 1.2 breaks XP compat, we can add a warning to the downloads page.

**Comment by musikBear** [comment]
> Please don't put up a big flashy warning on the forums.

just a simple post on support, has been posted.
http://lmms.io/forum/viewtopic.php?f=7&p=15790#p15790

**Comment by tresf** [comment]
> A guy with Vista also had problems.  http://lmms.io/forum/viewtopic.php?f=7&t=4822

His bug is ` i installed the beta version 1.1.9 in win vista and when i press edit/settings, it crashes every time , therefore it is impossible to use.`

That is a separate issue.  Please file a separate bug report.

> just a simple post on support, has been posted.  http://lmms.io/forum/viewtopic.php?f=7&p=15790#p15790

Your post title is:

> Important Info! Do NOT install 1.190 on XP!

Which is 1. Invalid, there's no such thing as 1.1.90.  2.  **Exactly** what was asked of you NOT to post.  Deleted.

> btw strange that lmms-1.1.1-g23d2824-win32.exe installed with no issues at all. Eg your build from january, was just fine. Runs perfect, also on xp

The `g23d2824` build was created using Ubuntu 12.04 on my own personal VM, whereas the version on our Downloads page was built using [Travis-CI](https://blog.travis-ci.com/2015-11-27-moving-to-a-more-elastic-future) so this may be a bug that we need to identify.  We'll test on some platforms and report back our findings.

> 1.1.9RC no entryPoint on KERNEL32

FYI, this version doesn't exist.  We appreciate the fast feedback as well as the strong enthusiasm, but the more the devs have to edit your posts and bug reports, the less time they have to fix problems. :+1:

**Comment by tresf** [comment]
1.2.0-RC1 Windows test Results:

Tests performed:  Installation, running.

| Version | OS | Status |
|-----------|-----|----------|
| win32 | Windows 7 x64 | :white_check_mark: |
| win32 | Windows XP SP3 | :no_entry_sign:  |
| win32 | Windows 10 x64 |  :white_check_mark: |
| win32 | Wine 32-bit | :hourglass: TBD |
| win64 | Windows 7 x64 | :white_check_mark: |
| win64| Windows 10 x64 |  :white_check_mark: |
| win32 | Wine 64-bit | :hourglass: TBD |


Picture, because everyone loves pictures... 
![image](https://cloud.githubusercontent.com/assets/6345473/13112620/89be018c-d559-11e5-975e-1ad1da65bcd3.png)

**Comment by tresf** [comment]
@musikBear we may be able to fix this, but we'll need someone a bit more familiar with the linking process to help.  Here's an `mingw-w64` article which I believe to be relevant to the message we're receiving:

https://sourceforge.net/p/mingw-w64/wiki2/Procedure%20entry%20point/

Tagging @Lukas-W @Umcaruje on a whim, but first we need to be able to reproduce the problem by building a win32 version on our own machines that suffers the bug, then we can make some attempts to fix this.  The downside is that [Wine doesn't have the same issue](https://github.com/LMMS/lmms/issues/2576#issuecomment-185240499), so we'll need to load up a 30-day XP VM to reproduce. https://dev.windows.com/en-us/microsoft-edge/tools/vms/windows/

In this case, I believe this to be a bug worth fixing.  Thanks for reporting.  P.S. Please upgrade your O.S. :beers:

**Comment by tresf** [comment]
Reproducible via `sudo apt-get upgrade` and clean build of `master`.

**Comment by musikBear** [comment]
@tresf Thanks! But this should not hold back development and or release. 'Us' with xp has to realize that the train has left the station for scrapping :+1:

**Comment by tresf** [comment]
@musikBear here you are: [`tresf/lmms-1.1.90-gd098a39-win32-nosoundio.exe`](https://github.com/tresf/lmms/releases/download/v1.2.0-RC1.2/lmms-1.1.90-gd098a39-win32-nosoundio.exe)

Ok... after using DependencyWalker, I was able to track down `libsoundio.dll` causing the crash.  In short, compiling with `-DWANT_SOUNDIO=False` will allow LMMS to continue running on Windows XP.

Turns out `soundio`'s [`InitOnceBeginInitialize`](https://github.com/andrewrk/libsoundio/blob/master/src/os.c#L582) is causing this crash on Windows XP, as it's a [Windows Vista and higher function](https://msdn.microsoft.com/en-us/library/windows/desktop/ms683487(v=vs.85).aspx).  The soundio page makes this requirement/limitation [very clear](https://cloud.githubusercontent.com/assets/6345473/13545939/2b5b3a14-e26b-11e5-9197-df5a52adc810.pn).

@andrewrk (and perhaps @LMMS/developers), I would like to know your thoughts on having `libsoundio` disabled for our win32 builds.
 * We don't have a whole lot of options for audio backends in Windows.  Has anyone tested this with LMMS? Specifically WASAPI vs. SDL?
 * If our website traffic is any indication, [Windows XP is usage is on a rapid decline](https://cloud.githubusercontent.com/assets/6345473/13545932/bd244694-e26a-11e5-9fa6-971d17e6f03f.png).


![image](https://cloud.githubusercontent.com/assets/6345473/13545896/23b4a220-e269-11e5-82b4-9b762b692a61.png)

**Comment by grejppi** [comment]
@tresf libsoundio apparently [requires Windows 7 or newer](http://libsound.io/)

**Comment by musikBear** [comment]
@tresf very nice! Thanks! :wine_glass: 
Perhaps this version could be the last farewell, with a informative note in the release doc, that hereafter winXP specifiv issues, will not be considered in future development. Then a nice clean sand-line has been drawn, with 1.2, and me and china will have to upgrade, or go linux.

**Comment by mikobuntu** [comment]
@musikBear just out of curiosity is XP still covered by security updates? And my 2 cents, I would recommend switching to Linux 😆

**Comment by IvanMaldonado** [comment]
Who uses vista anyway? Lmms was lucky to run on Vista since many many things didn't work properly even in Vista times.

**Comment by tresf** [comment]
I appreciate the feedback.

What I haven't read yet is a Windows 7/8/10 user that has tried libsoundio from our 1.2-RC1 release to vote in its favor.  We can pull this library from our 1.2 release and offer XP support for a bit longer, but I'm not sure which is worth more.

On that note, I'm attempting to compile libsoundio using Mingw so that I can experiment with [`GetProcessAddress` and family](https://sourceforge.net/p/mingw-w64/wiki2/Procedure%20entry%20point/) per MinGW documentation.

Mingw progress is here: https://github.com/andrewrk/libsoundio/issues/58

**Comment by tresf** [comment]
So I tried to use `libsoundio` today on Windows 10 64-bit, and these are my results:

 * Audio backends: 
  * `libsoundio: WASAPI (default device)`: Selectable, restart LMMS for changes to take effect.  Playback is too sporadic for daily use, but sounds ok when it does work.  Something seems to kill the audio randomly.  Note, this is not quite as bad as `PortAudio: Direct Sound`.  On a side note, `PortAudio: WASAPI` was working quite well.
  *  * `libsoundio: WASAPI (**raw** device)`: Selectable, restart LMMS for changes to take effect.  Software refuses to load after this option is selected.

So at this point, we can't ship `libsoundio` with Windows builds unless these sound quality and stability issues are addressed.

Shipping `libsoundio` based on sound quality and stability is really a topic for another thread, but I'll continue to leave this open until a decision is reached as it indirectly impacts the XP issues.

![image](https://cloud.githubusercontent.com/assets/6345473/13552390/c6b37706-e32e-11e5-94d1-69bb31b3ad22.png)

**Comment by musikBear** [comment]
@mikobuntu updates, no. I run some security stuff on top of xp. I consider that to be sufficient, and i have never had any incident. :four_leaf_clover: 

@tresf those are interesting, albeit i do have to say that ASIO4ALL imo is the best 'other than SDL' for windows. At current ASIO is a no-go, because of the mandatory crash-at-shutdown #2584. On my system that causes the BSOD every time -Eg ASIO causes a critical issue in windows, and BSOD flashes for a fraction of a second, and then the system reboot :angry: ..And *that* is totally unacceptable error handling, but so windows :-1: (an argument for your suggestion @mikobuntu :)
ASIO is imo the best sound, but that is ofcause only compared to SDL, because i have only horrible noise from any other setting.
Against ASIO is ofcause the fact that it cant be bundled at a<ll, and using it is in some way not legit :lollipop:

**Comment by tresf** [comment]
It is with mixed emotions that I say, we can continue using Windows XP on 1.2.0 per #2645. :smile: 

Reasoning had most to do with the unreliability of `libsoundio` + LMMS.

![screen shot 2016-03-07 at 12 55 10 am](https://cloud.githubusercontent.com/assets/6345473/13561829/7df7e20a-e3ff-11e5-99e5-487fdfc022ab.png)

**Comment by andrewrk** [comment]
Thanks for doing this testing. I'll see if I can figure out what's going wrong with libsoundio on Windows. Although I'll be testing on Windows 7 or Windows 10 and not XP.

**Comment by tresf** [comment]
>  and not XP.

Well, I got a [bit ambitious](https://github.com/andrewrk/libsoundio/commit/5f33f95) for a minute there, but quickly realized how much compat code would be needed just to prevent the Windows XP crash.  I'm still interested in a technique that allows it to silently fail, but I don't know enough about the compiler to tackle something like that.

> I'll be testing on Windows 7 or Windows 10 and not XP.

Thanks. We have issues with our PortAudio DirectSound driver too (PortAudio WASAPI seems to work) so our recommendation for years has been to use SDL, which seems to be quite high on CPU usage.  We'll be happy to re-enable `libsoundio` on Windows once it's working.  Much obliged.

**Comment by andrewrk** [comment]
> but quickly realized how much compat code would be needed just to prevent the Windows XP crash.

That's not even the real issue, the real issue is that libsoundio would need to target a different audio backend than WASAPI for XP.

**Comment by tresf** [comment]
> That's not even the real issue, the real issue is that libsoundio would need to target a different audio backend than WASAPI for XP.

This is where my understanding of the linker gets a bit foggy... by "different audio backend" you're essentially saying to ship a separate DLL just for XP, right?  Because everything I read about the linker said to either use soft-links against KERNEL32 to detect missing symbols, or don't run on that platform.  My attempt to find another way was unsuccessful. :beer:

**Comment by andrewrk** [comment]
I mean that I would need to write a DirectSound backend for libsoundio since WASAPI was introduced in Windows Vista.

**Comment by gwald** [comment]
lmms-1.1.90-win32.exe has the KERNEL32 missing issue for winXP.
I removed libsoundio.dll but it's still linked.
Where's the installer or lmms.exe for winXP?
Cheers.

**Comment by musikBear** [comment]
@gwald 
Here, if its still available, if not, tell here
https://github.com/LMMS/lmms/issues/2576#issuecomment-192584627

**Comment by gwald** [comment]
ha.. cool thanks, yes that works :D
